<a href="https://colab.research.google.com/github/Madhavi-sai/FML/blob/main/knn_on_diabetes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## KNN- K-Nearest Neighbours  
K-nearest neighbors (KNN) algorithm is a type of supervised ML algorithm which can be used for both classification as well as regression predictive problems. However, it is mainly used for classification predictive problems 



In [ ]:
#import the  necessary packages

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

Attach appropriate data set

In [ ]:
diabetes_data = pd.read_csv("/content/diabetes.csv")

In [ ]:
diabetes_data.head()

pregnancies  glucose  diastolic  triceps  insulin   bmi    dpf  age  \
0            6      148         72       35        0  33.6  0.627   50   
1            1       85         66       29        0  26.6  0.351   31   
2            8      183         64        0        0  23.3  0.672   32   
3            1       89         66       23       94  28.1  0.167   21   
4            0      137         40       35      168  43.1  2.288   33   

   diabetes  
0         1  
1         0  
2         1  
3         0  
4         1

In [ ]:
diabetes_data.columns

Index(['pregnancies', 'glucose', 'diastolic', 'triceps', 'insulin', 'bmi',
       'dpf', 'age', 'diabetes'],
      dtype='object')

In [ ]:
X = diabetes_data.drop("diabetes", axis = 1).values
y = diabetes_data['diabetes'].values

In [ ]:
print(type(X))
print(type(y))
print(X.shape, y.shape)

<class 'numpy.ndarray'>
<class 'numpy.ndarray'>
(768, 8) (768,)


From the above cell, we can conclude that there are 768 records, 
Our feature set consists of 8 columns namely "pregnancies', 'glucose', 'diastolic', 'triceps', 'insulin', 'bmi',
       'dpf', 'age'. We are going to label a person as diabetic or non-daibetic based upon these features.

In [ ]:
diabetes_data.diabetes.unique()

array([1, 0])

This is a binary classification problem, since the target column has two unique labels to classify upon.  
1 - The person is diabetic.  
0 - The person is non-diabetic.

Next, we will divide the data into train and test split. Following code will split the dataset into 80% training data and 20% of testing data −

In [ ]:
#split the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=41)

In [ ]:
X_train.shape, X_test.shape

((614, 8), (154, 8))

In [ ]:
y_train.shape, y_test.shape

((614,), (154,))

Data scaling will be done as follows −

In [ ]:
scaler = StandardScaler()
scaler.fit(X_train)
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

 Train the model with the help of KNeighborsClassifier class of sklearn as follows −

In [ ]:
#train the model
knn = KNeighborsClassifier(n_neighbors=1)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=1)

In [ ]:
y_pred = knn.predict(X_test)

In [ ]:
knn.score(X_train, y_train), knn.score(X_test, y_test)

(1.0, 0.7402597402597403)

In [ ]:
#train the model
knn = KNeighborsClassifier(n_neighbors=7)
knn.fit(X_train, y_train)
y_pred = knn.predict(X_test)
print(knn.score(X_train, y_train), knn.score(X_test, y_test))

0.7899022801302932 0.7792207792207793


## Model Performance Analysis

In [ ]:
confusion_matrix(y_test, y_pred)

array([[86, 13],
       [21, 34]])

The classification report is based upon the following evaluation measures.
**Accuracy** - The percentage of correctly labelled values out of all the values, from the confusion matrix.  
**Precision** - Precision is the ratio of correctly predicted positive observations to the total predicted positive observations.  
Precision = TP/TP+FP  
**Recall**- Recall is the ratio of correctly predicted positive observations to the all observations in actual class.  
Recall = TP/TP+FN   
**F1 score** - F1 Score is the weighted average of Precision and Recall. Therefore, this score takes both false positives and false negatives into account. 

In [ ]:
print(classification_report(y_test, y_pred))
print(accuracy_score(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.80      0.87      0.83        99
           1       0.72      0.62      0.67        55

    accuracy                           0.78       154
   macro avg       0.76      0.74      0.75       154
weighted avg       0.78      0.78      0.77       154

0.7792207792207793
